### DISCUSSION TITLE
Eurorack style Function Generators

### DISCUSSION THREAD

**Discussion by NoiseTheorem** [SynthstromAudible/DelugeFirmware#794] [Ideas]
For sound design purposes, I would like to see the addition of two Function Generators (FGs).   Each FG would have a Rise Time, Decay Time, and a shape (from exponential to linear to logarhythmic).  FGs would have three modes - OneShot, Triggered Loop, and Free Run.

**Comment by m-m-adams** [comment]
All the code exists for this; it just needs a UI mockup since I don't have good ideas on how to integrate it

**Comment by timriot** [comment]
Heyo! I'm interested in helping push this feature request along. I've got some time to put together a UI mockup, unless @NoiseTheorem or someone else has already done so. Are there any examples of UI mockups that I could use as a template or examples of the format that you'd suggest? I've already checked out the Deluge UX Principles documentation and the manual.

**Reply by sapphire-arches** [reply]
There's [a Figma template](https://www.figma.com/file/mySg4iPCuJUEUmbC251GnJ/Deluge?type=design&node-id=0%3A1&mode=design&t=zrpJLhfrG0F4YmL7-1), courtesy of Discord user laikrodiz. There's also some example OLED graphics [here](https://discord.com/channels/608916579421257728/1175903056240595074/1176254446846611546) thanks to ok-reza. The size is 43x128, in black and white.

**Reply by timriot** [reply]
Thank you @bobtwinkles. I'm familiarizing myself with all the templates and examples you shared. Will check back in once I've had a day to mull this over.

**Comment by timriot** [comment]
Here's the first draft! Please let me know what y'all think! I'm down to do any variations or revisions you'd like. Please let me know! FWIW, This is an extension of a system I'm starting to develop for the Deluge at large, which you can see in the Deluge OLED UI channel [here](https://discord.com/channels/608916579421257728/1175903056240595074/1192091822932119654).
<img width="661" alt="funcgen" src="https://github.com/SynthstromAudible/DelugeFirmware/assets/1869621/8f8077c7-3e1f-4e02-8669-b5ff45ad122a">

**Comment by timriot** [comment]
Here's a second draft that I put together after running into m-m-adams on Discord.

This tries to address the questions of – 
Where to put it on the grid and how you'd modulate with it? 
Would it be a mode for the ADSRs or LFOs (which one)? 
Live entirely in the menu, or take over some other shortcut?
<img width="705" alt="funcgen2" src="https://github.com/SynthstromAudible/DelugeFirmware/assets/1869621/8b521697-48ba-4c84-8e2c-a91193cb0ec9">

(Specifics of the mod I/O of the function generator still TBD, as it would be nice if it could mod different or more things than the standard ADSR Env or LFO.)

**Comment by timriot** [comment]
Here's a quick take on using the pads for some visualization / interaction. These aren't final, so please let me know if there's interested in developing these further and I'll put some more thinking in.
[Mods on pads - Deets.pdf](https://github.com/SynthstromAudible/DelugeFirmware/files/13852343/Mods.on.pads.-.Deets.pdf)

**Reply by timriot** [reply]
<img width="906" alt="mods_on_pads" src="https://github.com/SynthstromAudible/DelugeFirmware/assets/1869621/a5567b82-3603-4834-b31a-31aca8a64b25">
